discojs/src/models/gpt/layers.spec.ts

```diff
@@ -0,0 +1,308 @@
+    });
+
+    it('should throw an error if input is an array with more than one tensor', function () {
+      const vocabSize = 100;
+      const nEmbd = 16;
+      const seed = 42;
+      const lmEmbedding = new LMEmbedding(vocabSize, nEmbd, seed);
+      const input1 = tf.tensor2d([[1, 2, 3]], [1, 3], 'int32');
+      const input2 = tf.tensor2d([[4, 5, 6]], [1, 3], 'int32');
+      expect(() => lmEmbedding.apply([input1, input2])).to.throw('expected exactly one tensor');
+    });
+
+    it('should compute correct output shape for 2D input using computeOutputShape', function () {
+      const vocabSize = 100;
+      const nEmbd = 16;
+      const seed = 42;
+      const lmEmbedding = new LMEmbedding(vocabSize, nEmbd, seed);
+      const outputShape = lmEmbedding.computeOutputShape([null, null]);
+      expect(outputShape).to.deep.equal([null, null, nEmbd]);
+    });
+
+  });
+
+  // Range Layer tests
+  describe('Range Layer', function () {  
+
+    afterEach(() => {
+      // dispose any created tensors/variables
+      tf.disposeVariables();
+    });
+  
+    it('should output a tensor with shape [1, T] for an input of shape [batch, T]', async function () {
+      const rangeLayer = new Range();
+  
+      // dummy input tensor with shape [batch, T]
+      const dummyInput = tf.zeros([3, 10], 'int32');
+  
+      const output = rangeLayer.apply(dummyInput) as tf.Tensor;
+  
+      // We expect the output to have shape [1, T] i.e. [1, 10]
+      expect(output.shape).to.deep.equal([1, 10]);
+  
+      // verify the content: the layer should output a range [0, 1, ..., T-1]
+      const outputData = await output.data();
+      for (let i = 0; i < 10; i++) {
+        expect(outputData[i]).to.equal(i);
+      }
+    });
+  });
+
+  // MLP Layer tests
+  describe('MLP Layer', function () {  
+
+    it('should produce deterministic/non-NaN outputs with the same random seed', async function () {
+      // an MLP config with a fixed seed
+      const config: MLPConfig = {
+        name: 'testMLP',
+        contextLength: 10,
+        residDrop: 0,  // no dropout for deterministic behavior
+        nLayer: 2,
+        seed: 42,
+        nEmbd: 16,
+        nHead: 4
+      };
+  
+      // two separate MLP model instances using the same config
+      const model1 = MLP(config);
+      const model2 = MLP(config);
+  
+      const input = tf.ones([1, config.contextLength, config.nEmbd]);
+  
+      // get predictions from both models
+      const output1 = model1.predict(input) as tf.Tensor;
+      const output2 = model2.predict(input) as tf.Tensor;
+  
+      const arr1 = await output1.data();
+      const arr2 = await output2.data();
+  
+      // check lengths are equal
+      expect(arr1.length).to.equal(arr2.length);
+  
+      // check that the models produce the same output
+      expect(arr1).to.deep.equal(arr2);
+
+      // Check that there are no NaN values in the outputs.
+      for (let i = 0; i < arr1.length; i++) {
```

Comment:
better to use `for … of` to loop over an array
```suggestion
      for (const v of arr1) {
```